src/connection/jmx.go

```diff
@@ -0,0 +1,102 @@
+package connection
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/newrelic/nri-kafka/src/args"
+	"golang.org/x/sync/semaphore"
+	"sync"
+
+	"github.com/newrelic/nrjmx/gojmx"
+)
+
+var (
+	ErrJMXCollection = errors.New("JMX collection failed")
+	ErrConnectionErr = errors.New("JMX connection failed")
+
+	jmxProvider JMXProvider
+	once        sync.Once
+)
+
+type JMXProvider interface {
+	NewConnection(config *gojmx.JMXConfig) (conn JMXConnection, err error)
+}
+
+// JMXProviderWithLimit will be able to allocate new JMX connections, as long as maxConnections is not reached.
+type JMXProviderWithLimit struct {
+	ctx context.Context
+	sem *semaphore.Weighted
+}
+
+func GetJMXConnectionProvider() JMXProvider {
+	once.Do(func() {
+		jmxProvider = newJMXProviderWithLimit(context.Background(), args.GlobalArgs.MaxJMXConnections)
+	})
+	return jmxProvider
+}
+
+func SetJMXConnectionProvider(provider JMXProvider) {
+	once.Do(func() {})
+	jmxProvider = provider
+}
+
+// newJMXProviderWithLimit creates a new instance of JMXProvider.
+func newJMXProviderWithLimit(ctx context.Context, maxConnections int) JMXProvider {
+	return &JMXProviderWithLimit{
+		ctx: ctx,
+		sem: semaphore.NewWeighted(int64(maxConnections)),
+	}
+}
+
+// NewConnection will return a new JMX connection if the maximum number of concurrent connections
+// is not reached, otherwise will block until a connection is released.
+func (p *JMXProviderWithLimit) NewConnection(config *gojmx.JMXConfig) (conn JMXConnection, err error) {
+	err = p.sem.Acquire(p.ctx, 1)
+	if err != nil {
+		err = fmt.Errorf("failed to open new connection, %w", err)
+		return
+	}
+
+	defer func() {
+		if err != nil {
+			p.sem.Release(1)
+		}
+	}()
```

Comment:
I think I get what you want to achieve here (release after this func is called if there's an error) but I think doing it this way is not very intuitive for the reader.

Is there a different way in which we could do this? e.g. locking the semaphore after we check that `gojmx.NewClient` returns successfully?